package/MDAnalysis/lib/_cutil.pyx

```diff
@@ -186,14 +187,14 @@ def make_whole(atomgroup, reference_atom=None, inplace=True):
         are returned as a numpy array.
     """
     cdef intset refpoints, todo, done
-    cdef int i, j, nloops, ref, atom, other, natoms
+    cdef np.intp_t i, j, nloops, ref, atom, other, natoms
```

Comment:
I don't this is going to break anything, but fwiw the `intmap` above uses `int` types, whereas this becomes a `ssize_t` type.  I think `np.intc` is the better match for `int` in C

https://docs.scipy.org/doc/numpy/user/basics.types.html#array-types-and-conversions-between-types

Can change the `map`/`set` objects up top to use the indexing integers and everyone is happy?

Reply:
I just changed this to get rid of compiler warnings. `ssize_t` will translate to `int32_t` on 32 bit systems and to `int64_t` on 64 bit machines. It cannot break anything since it is downcast when assigning it to an intmap.
`np.intc` is a pretty bad choice for counter variables since it translates to `int` in C.  That again is very often identical to `int32_t` but the standard only guarantees at least 16 bits!

Speaking of integer widths, there is something quite odd in MDAnalysis, which may become a problem in the (probably *very* distant) future: Atom indices are stored as `np.int64` and bond indices, which again refer to atoms, are stored as `np.int32`. That doesn;t really make sense, right?

Reply:
Yeah sounds like both should be 32 or 64. I doubt we're going to need 64 bit ints to represent them? Or better to have 64bit ints so we can quickly index using atom indices (common)?

Reply:
I think 32bits should be way enough to accomodate even giant system sizes. 2 ** 31 - 1 is pretty large after all.
IIRC numpy uses `np.intp` for indexing, so one should choose that for index arrays. (unless someone runs MDA on a 16 bit machine... :stuck_out_tongue_winking_eye:)

Reply:
I just tried to correct all the types in `_cutil.pyx` but it turns out to be in fact impossible. The reason is exactly the difference between the atom index representation in `AtomGroup._ix` (`np.int64`) and the bond indicex representation (`np.int32`).

I wanted to find out which type is actually used to store the bond indices and worked my way through the entirely messed up `Bonds`/`_Connection`/`TopologyAttr`/`TopologyGroup` zoo:

`type(u.atoms.bonds)` says it's a `TopologyGroup`, so that's where I started.
`TopologyGroup.indices` of an empty `TopologyGroup` returns an empty `np.int32` array even though its `_bix` member is default-initialized with `np.array([])`, which defaults to `np.float64`.
A non-empty `TopologyGroup` will return `TopologyGroup._bix` when `TopologyGroup.indices` is accessed. `TopologyGroup._bix` has the dtype of whatever is passed as its `bondidx` argument in `TopologyGroup.__init__()`.
Such a `TopologyGroup` object is returned from a `_Connection` object (`Bonds` is one) in its `_Connection.get_atoms()` method. This method initializes the `TopologyGroup` with an `np.int32` index array which is compiled from `_Connection`'s cached `_bondDict` member. The indices stored in `_Connection._bondDict` have the dtype of whatever was used to initialize `_Connection.values` from the `values` argument of `_Connection.__init__()`.
If I got it right, `_Connection` objects usually come into life when a `Universe` is created, and are effectively instantiated by either a topology parser,  in `core.Universe.Merge()`, or in `AtomGroup.guess_bonds()`.
I really didn't want to check all topology parsers for the kinds of bond objects they return (quickly grepping the parsers suggests mostly `np.int32` but also sometimes `np.int64` index arrays for atoms and/or bonds), but I could figure out that `core.Universe.Merge()` instantiates bonds with a list of `(np.int64, int)` tuples (yes, the second item is a Python int) as the `values` argument.
`AtomGroup.guess_bonds()` boils down to initializing bonds with a list of `(np.int64, np.int64)` tuples (atom indices) in `topology.guessers.guess_bonds()`.

So, in the end, I guess it's (mostly, but, as we could see, not always) up to the topology parser to choose the data type used to store the *actual* bond indices.

Now, what happens if one calls `atoms.bonds.to_indices()`? That's where the journey became quite interesting:
`atoms.bonds` invokes the "injected" getter method `TopologyAttr.__getitem__()`, which then invokes `self.get_atoms(atoms)`, which is implemented in  `_Connection.get_atoms(atoms)`. In there, a `TopologyObject` is instantiated with an `np.int32` index array created from whatever resides in `self._bondDict` (discussed above).
Now, since `atoms.bonds` boils down to an *always newly created* `TopologyObject`, calling `atoms.bonds.to_indices()` means actually calling that object's `to_indices()` method, which again calls its own `indices` property, which then returns unique elements of the very same index array it had just been instantiated with.

I wouldn't even dare to mention all the overhead taking place in the various methods and initializers on the way. I always wondered why working with these things is so damn slow. Now I know.

Reply:
Long story short, changing the bond index dtype to `np.int64` or, ideally, to `np.intp` is a piece of cake. However, changing the dtype of atom indices to `np.intp` will break a lot of low-level thingies implemented on the Cython- or C-level.

Reply:
@zemanj I'm happy to admit the workings of bonds can be improved, it was one of the first things I ever contributed to MDA :p And when I rewrote how topologies worked I wasn't really concentrating on getting bonds "nice".  That might be why there's a confusing 2 layer feel for how they work, there's essentially the `TopologyGroup` system I did first, then this got translated into a `TopologyAttribute` quite lazily (as I was putting out fires in other places).

The `dtype` of a TopologyAttribute should* be defined in the class (there's an `asarray` call to enforce this), so it can be singly defined there, then parsers should be hands off wrt defining this.

I think maybe we can squash some of the bond functionality (eg `TopologyDict`, maybe an "everything" `TopologyGroup` a la `u.atoms`) into the `_Connection` class.

* there's going to be a lot of shoulds that aren't followed, but that's just required cleanup

Reply:
@richardjgowers I am sorry, my judgement was probably a bit too harsh. I was just a bit frustrated as you might have imagined. :wink: 
I think there's always going to be room for improvement, and that certainly also applies to my own contributions.

In the future, we should make sure to use (global) index arrays instead of lists/tuples/dicts of objects whenever that's feasible, i.e., get as close as possible to using primitive types. Of course, one should also have maintainability in mind, so there might be some trade-offs to consider.